### ISSUE TITLE
about Loss Function

### ISSUE THREAD

**Issue by SIAAAAAA** [aurora95/Keras-FCN#13]
HI, I'm new in this field, I use your code to do segmentation. I confuse about the loss func. 
For example I have two classes. [0] represents background class and [1] represent another class. My label '.png' has the shape like [0 0 0 0 1 1 0 0...] (2dim). Now I use `softmax_sparse_crossentropy_ignoring_last_label(y_true, y_pred)` as my loss function.

It says 

> ignoring_last_label

I'm wondering which label it will ignore? class[0], class[1], or..another?
I trace the code and find ..it seems still have 2 classes shape.

my y_pred, the out put of my model has the shape (num,224,224,2)
And I want to calculate 2 classes loss.

Many Thanks!

**Comment by aurora95** [comment]
Well it's an ugly implementation actually... The ignoring_last_label is used for many segmentation datasets like PASCAL VOC or Cityscapes, since they have a 255 label denoting the edge of objects or unknown part that should be ignored. I change 255 to nb_classes in data generator so that it becomes the last label, and then I ignore it in the loss function. 
For your case, I think you can simply use a softmax of sigmoid loss (maybe you want to wrap a tensorflow loss instead of directly using the losses from keras, i'm not sure).

**Comment by SIAAAAAA** [comment]
Thank you for your response^^.
I save img&label generated by ` SegDataGenerator` I find that the `label_cval,` set 255. The label'.png' has white background around. So maybe it ignore this label? I'm not sure ....
Also my label'1' area is too small compared with '0'. I'm wondering if it is possible to multiply different weights to different label then calculate loss, changing some parts of the loss function.

**Comment by aurora95** [comment]
Yes by default ignore_label and label_cval are both 255 and it will be ignored.

You can pass a class_weight to fit_generator, note that considering the ignored label, there is actually nb_classes+1 classes. But some paper reports that for segmentation you don't need to deal with the unbalanced samples and it can still get good results, so maybe you can just try not using class_weights first.

**Comment by Ssica** [comment]
Hi @aurora95,
I have a quick question regarding your loss function too I was hoping you could answer:
I have my training data: 155x240x4 and my labels: 155x240 (a segmented image), am I still able to use your loss function? I'm unsure because of the following line in your loss function:
                   y_true = K.one_hot(tf.to_int32(K.flatten(y_true)), K.int_shape(y_pred)[-1]+1)
where u create a one_hot matrix from your y_train.

**Comment by aurora95** [comment]
@Ssica Yes you should be able to use the loss, and yes it's the code that removes the last label

**Comment by godissaw** [comment]
Hi @aurora95 
use your loss and accurate funcion, we can fit the model.
But when I want to use the model to predict, the output has nb_classes maps, what should I do?
should I add softmax activation layer then use the argmax?
I'm sorry to bother you ,but I hope that you can give me some short codes to show how to make the (21,w,d) into (1, w, d) martix and whether we need softmax, because i had some issues when I try to implement it. Thanks a lot.

**Comment by aurora95** [comment]
@godissaw You can directly use argmax, since argmax(y) would yield the same result with argmax(softmax(y))
